docs/developer/design-v2-2023-11-11.md

```diff
@@ -0,0 +1,198 @@
+# Design Document for Materialize Terraform Provider
+
+The proposed updates to the Materialize Terraform provider aim to streamline the management of resources within Materialize, offering a unified interface for global and regional resources. The provider will leverage a single authentication method, an app password, to derive the necessary configurations for seamless interaction with various services. It will also incorporate a dynamic client allocation mechanism to handle different resource types and their requirements.
+
+## Provider Overview
+
+### Objectives
+
+-   To provide a singular Terraform provider capable of interfacing with multiple backends and services.
+-   To facilitate the management of global resources and regional deployments through a consistent workflow.
+-   To enable a high degree of automation in resource provisioning and management, thereby reducing manual overhead and potential for errors.
+
+### Features
+
+-   **Single Authentication**: Utilizing an app password to access and manage resources without the need for multiple credentials.
+-   **Global and Regional Resource Management**: Allowing users to manage resources that are not bound by regional constraints, as well as those that are region-specific.
+-   **Dynamic Client Allocation**: Depending on the resource's requirement, the appropriate client (DB, Frontegg, Cloud API and etc.) will be allocated dynamically.
+
+## Architecture
+
+### Provider Configuration
+
+The provider will be initialized with parameters that determine its behavior and default settings:
+
+```hcl
+provider "materialize" {
+  app_password   = var.materialize_app_password
+  default_region = var.materialize_default_region
+  endpoint       = var.materialize_endpoint
```

Comment:
We should hide this one from users.

Reply:
As far as I can tell the Terraform's official documentation doesn't explicitly detail best practices for hiding provider configuration options. Probably because typically, all configuration options are intended to be exposed to users.

Maybe an option here is to remove the endpoint from the provider configuration schema, and only expose it as an `MZ_ENDPOINT` env variable. We can document that in the dev docs. Open to suggestions @dehume

Reply:
Ah yeah I was imagining it would work like the `sslmode` option today, where it's available as a configuration option but we just hide it in the docs. Only supporting it via the `MZ_ENDPOINT` environment variable sounds like a reasonable option too.

Reply:
What Nikhil said. We entirely override the [main provider page](https://github.com/MaterializeInc/terraform-provider-materialize/blob/main/docs/index.md). So it would not be immediately exposed to users unless we add it.

Reply:
Sounds good! I will keep it as the `sslmode` option in this case!